op-superchain/backend.go

```diff
@@ -0,0 +1,147 @@
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+type Backend interface {
+	MessageSafety(context.Context, MessageIdentifier, hexutil.Bytes) (MessageSafetyLabel, error)
+}
+
+type backend struct {
+	log log.Logger
+	mu  sync.RWMutex
+
+	l2FinalizedHeadSub  ethereum.Subscription
+	l2FinalizedBlockRef *eth.L1BlockRef
+
+	l2PeerNodes map[uint64]client.RPC
+}
+
+func NewBackend(ctx context.Context, log log.Logger, m metrics.Factory, cfg *Config) (Backend, error) {
+	log = log.New("module", "superchain")
+	backend := backend{log: log, l2PeerNodes: map[uint64]client.RPC{}}
+
+	rpcOpts := []client.RPCOption{client.WithDialBackoff(10)}
+	l2Node, err := client.NewRPC(ctx, log, cfg.L2NodeAddr, rpcOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to L2 node: %w", err)
+	}
+
+	for chainId, l2NodeAddr := range cfg.PeerL2NodeAddrs {
+		l2Node, err := client.NewRPC(ctx, log, l2NodeAddr, rpcOpts...)
+		if err != nil {
+			return nil, fmt.Errorf("failed to connect to Peer L2 node, %d: %w", chainId, err)
+		}
+		backend.l2PeerNodes[chainId] = l2Node
+	}
+
+	// retrieve the current references before setting up the poll
+	l2BlockRefsClient := &blockRefsClient{l2Node}
+	finalizedHeadRef, err := l2BlockRefsClient.L1BlockRefByLabel(ctx, eth.Finalized)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query finalized block ref: %w", err)
+	}
+
+	backend.l2FinalizedBlockRef = &finalizedHeadRef
+	log.Info("detected finalized head", "number", finalizedHeadRef.Number, "hash", finalizedHeadRef.Hash)
+
+	l2FinalizedHeadSignal := func(ctx context.Context, sig eth.L1BlockRef) {
+		log.Info("new finalized head", "number", sig.Number, "hash", sig.Hash)
+		backend.mu.Lock()
+		backend.l2FinalizedBlockRef = &sig
+		backend.mu.Unlock()
+	}
+
+	pollInterval, timeout := time.Second, time.Second*10
+	backend.l2FinalizedHeadSub = eth.PollBlockChanges(log, l2BlockRefsClient, l2FinalizedHeadSignal, eth.Finalized, pollInterval, timeout)
+
+	return &backend, nil
+}
+
+func (b *backend) MessageSafety(ctx context.Context, id MessageIdentifier, payloadBytes hexutil.Bytes) (MessageSafetyLabel, error) {
+	b.log.Info("message safety check", "chain_id", id.ChainId, "block_number", id.BlockNumber, "log_index", id.LogIndex)
+
+	// ChainID Invariant.
+	//   TODO: Assumption here that the configured peers exactly maps to the registered dependency set.
+	//   When the predeploy is added, this needs to be tied to the dependency set registered on-chain
+	//   TODO: Either assume chain id never exceeds uint64 or handle this appropriately
+	l2Node, ok := b.l2PeerNodes[id.ChainId.Uint64()]
+	if !ok {
+		return MessageUnknown, fmt.Errorf("peer with chain id %d is not configured", id.ChainId)
+	}
+
+	var logs []types.Log
+	var header *types.Header
+
+	// Since eth_getLogs doesn't support specifying the log index, we fetch
+	// all the outbox reciepts for this block (TODO: add address filter). The
```

Comment:
Can we create a higher level abstraction to encapsulate this low level complexity?

Reply:
Moved out the log check into it's own function. Are you also thinking about an abstraction for fetching the receipts & info?

Reply:
Also pulled out the batch request into a LogsProvider abstraction.

Ideally we'd be able to use the l2_client within `op-service/sources` but that requires making some changes we can do later.